File: darwin/hd_sentry/Sources/hd_sentry/HdSentryBreadcrumbStore.swift
import Foundation

/// Session breadcrumbs persisted beside crash reports. Used by [HdSentryCrashStore] and other
/// native code in the host app (requires [HdSentryCrashStore.configure] first).
public enum HdSentryBreadcrumbStore {
  public static let breadcrumbFileName = "session_breadcrumbs.log"
  public static let maxBreadcrumbs = 100
  private static let fieldSep = "\t"

  public static func breadcrumbFileURL(in crashDirectory: URL) -> URL {
    crashDirectory.appendingPathComponent(breadcrumbFileName)
  }

  public static func add(
    crashDirectory: URL,
    message: String,
    category: String? = nil,
    data: String? = nil
  ) throws {
    try FileManager.default.createDirectory(at: crashDirectory, withIntermediateDirectories: true)
    let fileURL = breadcrumbFileURL(in: crashDirectory)
    let line = formatLine(message: message, category: category, data: data)
    if FileManager.default.fileExists(atPath: fileURL.path) {
      let handle = try FileHandle(forWritingTo: fileURL)
      defer { try? handle.close() }
      try handle.seekToEnd()
      if let data = line.data(using: .utf8) {
        try handle.write(contentsOf: data)
      }
    } else {
      try line.write(to: fileURL, atomically: true, encoding: .utf8)
    }
    try trimToMaxLines(fileURL: fileURL)
  }

  public static func clear(crashDirectory: URL) throws {
    let fileURL = breadcrumbFileURL(in: crashDirectory)
    if FileManager.default.fileExists(atPath: fileURL.path) {
      try FileManager.default.removeItem(at: fileURL)
    }
  }

  /// Formatted `--- breadcrumbs ---` section for a crash report; clears the log after read.
  public static func consumeFormattedSection(crashDirectory: URL) -> String {
    let fileURL = breadcrumbFileURL(in: crashDirectory)
    guard FileManager.default.fileExists(atPath: fileURL.path),
      let raw = try? String(contentsOf: fileURL, encoding: .utf8)
    else {
      return ""
    }
    try? FileManager.default.removeItem(at: fileURL)
    let lines = raw.split(whereSeparator: \.isNewline).map(String.init).filter { !$0.isEmpty }
    guard !lines.isEmpty else { return "" }
    var body = "\n--- breadcrumbs ---\n"
    for line in lines {
      body += line + "\n"
    }
    return body
  }

  private static func formatLine(message: String, category: String?, data: String?) -> String {
    let formatter = ISO8601DateFormatter()
    formatter.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
    let timestamp = formatter.string(from: Date())
    let cat = sanitizeField(category ?? "")
    let msg = sanitizeField(message)
    let extra = sanitizeField(data ?? "")
    return "\(timestamp)\(fieldSep)\(cat)\(fieldSep)\(msg)\(fieldSep)\(extra)\n"
  }

  private static func sanitizeField(_ value: String) -> String {
    value
      .replacingOccurrences(of: "\t", with: " ")
      .replacingOccurrences(of: "\n", with: " ")
      .replacingOccurrences(of: "\r", with: " ")
  }

  private static func trimToMaxLines(fileURL: URL) throws {
    guard let raw = try? String(contentsOf: fileURL, encoding: .utf8) else { return }
    var lines = raw.split(whereSeparator: \.isNewline).map(String.init)
    if lines.count <= maxBreadcrumbs { return }
    lines = Array(lines.suffix(maxBreadcrumbs))
    try lines.joined(separator: "\n").appending("\n").write(to: fileURL, atomically: true, encoding: .utf8)
  }
}

File: android/src/main/kotlin/com/hodoan/hd_sentry/HdSentryBreadcrumbStore.kt
package com.hodoan.hd_sentry

import java.io.File
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale
import java.util.TimeZone

/**
 * Session breadcrumbs persisted beside crash reports. Shared by [HdSentryCrashStore] and
 * other native code in the host app (call after [HdSentryCrashStore.configure]).
 */
object HdSentryBreadcrumbStore {
    const val BREADCRUMB_FILE_NAME = "session_breadcrumbs.log"
    const val MAX_BREADCRUMBS = 100
    private const val FIELD_SEP = '\t'

    fun breadcrumbFile(crashDirectory: File): File =
        File(crashDirectory, BREADCRUMB_FILE_NAME)

    fun add(
        crashDirectory: File,
        message: String,
        category: String? = null,
        data: String? = null,
    ) {
        crashDirectory.mkdirs()
        val file = breadcrumbFile(crashDirectory)
        val line = formatLine(message, category, data)
        file.appendText(line)
        trimToMaxLines(file)
    }

    fun clear(crashDirectory: File) {
        breadcrumbFile(crashDirectory).delete()
    }

    /** Formatted `--- breadcrumbs ---` section for a crash report; clears the log after read. */
    fun consumeFormattedSection(crashDirectory: File): String {
        val file = breadcrumbFile(crashDirectory)
        if (!file.isFile) return ""
        val lines = file.readLines().filter { it.isNotBlank() }
        file.delete()
        if (lines.isEmpty()) return ""
        return buildString {
            appendLine()
            appendLine("--- breadcrumbs ---")
            lines.forEach { appendLine(it) }
        }
    }

    private fun formatLine(message: String, category: String?, data: String?): String {
        val timestamp = isoTimestamp()
        val cat = sanitizeField(category.orEmpty())
        val msg = sanitizeField(message)
        val extra = sanitizeField(data.orEmpty())
        return "$timestamp$FIELD_SEP$cat$FIELD_SEP$msg$FIELD_SEP$extra\n"
    }

    private fun sanitizeField(value: String): String =
        value.replace('\t', ' ').replace('\n', ' ').replace('\r', ' ')

    private fun trimToMaxLines(file: File) {
        val lines = file.readLines()
        if (lines.size <= MAX_BREADCRUMBS) return
        file.writeText(lines.takeLast(MAX_BREADCRUMBS).joinToString("\n", postfix = "\n"))
    }

    private fun isoTimestamp(): String {
        val formatter = SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'", Locale.US)
        formatter.timeZone = TimeZone.getTimeZone("UTC")
        return formatter.format(Date())
    }
}
